File: src/hardware/Screen.java
package hardware;

import java.awt.*;
import java.net.URL;
import java.awt.image.*;

public class Screen extends Canvas {

    BufferedImage BuffImg;
    WritableRaster raster;
    int[] pixels;
    protected Memory mem;
    private int pixelSize;
    private Graphics og;
    public boolean filter = false;
    private final int palette[] = {
        0x000000,
        0xFF0000,
        0x00FF00,
        0xFFFF00,

        0x0000FF,
        0xFF00FF,
        0x00FFFF,
        0xFFFFFF,

        0x636363,
        0xFF6363,
        0x63FF63,
        0xFFFF63,

        0x0063FF,
        0xFF63FF,
        0x63FFFF,
        0xFF6300,
    };

    public static int led = 0;
    public static int show_led = 0;
    final URL monito = getClass().getResource("mask.png");
    final Image monitor = getToolkit().getImage(monito);

    public Screen() {
        this.pixelSize = 2;
        BuffImg = new BufferedImage(320, 200, BufferedImage.TYPE_INT_RGB);
        raster = BuffImg.getRaster();
        pixels = new int[320 * 200];
        for (int i = 0; i < pixels.length; i++) {
            pixels[i] = 0xff000000;
        }
        BuffImg.setData(raster);
    }

    public void init(Memory memory) {
        this.mem = memory;
    }

    public void setPixelSize(int ps) {
        pixelSize = ps;
        mem.setAllDirty();
    }

    public int getPixelSize() {
        return pixelSize;
    }

    public void update(Graphics gc) {
        paint(gc);
    }

    public void paint(Graphics gc) {
        raster.setDataElements(0, 0, 320, 200, pixels);
        if (filter) {
            Graphics2D g2 = (Graphics2D) gc;
            g2.setRenderingHint(RenderingHints.KEY_INTERPOLATION,
                    RenderingHints.VALUE_INTERPOLATION_BILINEAR);
            gc = g2;
        }
            og = BuffImg.getGraphics();
            if (show_led > 0) {
                show_led--;
                if (led != 0)
                og.setColor(Color.red);
                else
                og.setColor(Color.BLACK);
                og.fillRect(320 - 16, 0, 16, 8);
            }
        dopaint(og);
        if (filter)
            og.drawImage(monitor, 0, 0, 320,200,this);
        gc.drawImage(BuffImg, 0, 0, 320 * pixelSize, 200 * pixelSize, this);
    }
    int i;
    int j;
    int xPosition;
    int yPosition;
    int col;
    int pt;
    int c1;
    int c2;
    int w;
    int p;
    int xx;
    int x;
    int y;
    int offset;

    public void dopaint(Graphics gc) {
        p = 0;
        i = 0;

        for (y = 0; y < 200; y++) {
            offset = y*320;
            x = 0;
            if (!mem.isDirty(y)) {
                i += 40;
            } else {
                for (xx = 0; xx < 40; xx++) {
                    int cc1;
                    int cc2;
                    col = mem.COLOR(i);
                    c2 = col & 0x0F;
                    c1 = col >> 4;
                    cc2 = palette[c1];
                    cc1 = palette[c2];

                    pt = mem.POINT(i);
                    if ((0x80 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    if ((0x40 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    if ((0x20 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    if ((0x10 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    if ((0x08 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    if ((0x04 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    if ((0x02 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    if ((0x01 & pt) != 0) {
                        pixels[x+offset] = cc2;
                    } else {
                        pixels[x+offset] = cc1;
                    }
                    x++;
                    i++;
                }
            }
        }


    }
}

File: src/hardware/screen.rs
use crate::domension::Dimension;
use crate::hardware::memory::Memory;
use crate::raw_image::RawImage;
use crate::{int, raw_image};
use std::cmp;

const PALETTE: [u32; 16] = [
    0x000000, 0xF00000, 0x00F000, 0xF0F000, 0x0000F0, 0xF000F0, 0x00F0F0, 0xF0F0F0, 0x636363,
    0xF06363, 0x63F063, 0xF0F063, 0x0063F0, 0xF063F0, 0x63F0F0, 0xF06300,
];

pub(crate) const WIDTH: usize = 320;
pub(crate) const HEIGHT: usize = 200;
pub(crate) const DEFAULT_PIXEL_SIZE: usize = 3;

#[derive(Debug)]
pub(crate) struct Screen {
    pub(crate) mouse_clic: bool,
    pub(crate) mouse_x: int,
    pub(crate) mouse_y: int,
    pub(crate) pixels: Vec<u32>,
    pub(crate) rgb_pixels: Vec<u8>,
    filter: bool,
    pub(crate) led: int,
    pub(crate) show_led: int,
    ratio: usize,
}

impl Screen {
    pub(crate) fn new() -> Self {
        Screen {
            mouse_clic: false,
            mouse_x: -1,
            mouse_y: -1,
            pixels: vec![0xff000000; WIDTH * HEIGHT],
            rgb_pixels: Vec::new(),
            filter: false,
            led: 0,
            show_led: 0,
            ratio: DEFAULT_PIXEL_SIZE,
        }
    }

    pub(crate) fn new_size(&mut self, new_size: Dimension) {
        let x_ratio = new_size.width / WIDTH;
        let y_ratio = new_size.height / HEIGHT;
        self.ratio = cmp::min(x_ratio, y_ratio);
    }

    pub(crate) fn paint(&mut self, mem: &mut Memory) {
        if self.show_led > 0 {
            //todo : restore this
            // self.show_led -= 1;
            // let color = if self.led != 0 {
            //     Color::from_rgb(255., 0., 0.)
            // } else {
            //     Color::from_rgb(0., 0., 0.)
            // };
            // let rectangle: Rectangle<f32> = Rectangle::new(Vector2::new(WIDTH as f32 - 16., 0.), Vector2::new(16., 8.));
            //  graphics.draw_rectangle(rectangle, color);
        }
        self.dopaint(mem);
    }

    pub(crate) fn get_pixels(&self) -> RawImage {
        let pixel_size = self.ratio;
        let mut raw_image = RawImage::new(WIDTH * pixel_size, HEIGHT * pixel_size);

        let buffer = &mut raw_image.data;
        let mut index = 0;
        for y in 0..HEIGHT {
            for _ in 0..pixel_size {
                for x in 0..WIDTH {
                    for _ in 0..pixel_size {
                        let p = self.pixels[x + y * WIDTH];
                        let r = (p & 0xFF) as u8;
                        let g = ((p >> 8) & 0xFF) as u8;
                        let b = ((p >> 16) & 0xFF) as u8;
                        buffer[index] = b;
                        index += 1;
                        buffer[index] = g;
                        index += 1;
                        buffer[index] = r;
                        index += 1;
                    }
                }
            }
        }

        raw_image
    }

    pub(crate) fn dopaint(&mut self, mem: &mut Memory) {
        let mut i = 0;

        for y in 0..HEIGHT {
            let offset: usize = y * WIDTH;
            if !mem.is_dirty(y) {
                i += 40;
            } else {
                let mut x: usize = 0;
                for _ in 0..40 {
                    let col = mem.COLOR(i);
                    let c2 = (col & 0x0F) as usize;
                    let c1 = (col >> 4) as usize;
                    let cc2 = PALETTE[c1];
                    let cc1 = PALETTE[c2];

                    let pt = mem.POINT(i);
                    const PATTERN: [int; 8] = [0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01];
                    for v in PATTERN {
                        if (v & pt) != 0 {
                            self.pixels[x + offset] = cc2;
                        } else {
                            self.pixels[x + offset] = cc1;
                        }
                        x += 1;
                    }
                    i += 1;
                }
            }
        }
    }
}
